charmcraft/main.py

```diff
@@ -58,41 +86,89 @@ def _get_option(args, name):
-        for ag in self._action_groups:
-            if ag.title is self.special_group:
-                self._action_groups.remove(ag)
-                main = True
-                break
-        base = super().format_help()
-        if not main:
-            return base
-
-        # Get the size of the longest name so all help texts are aligned
-        # properly across the groups.
-        longest_name = 0
-        for _, _, cmd_classes in self.__commands_groups:
-            for cmd_class in cmd_classes:
-                longest_name = max(len(cmd_class.name), longest_name)
-
-        extra = ['', 'commands:', '']
-        for group, group_title, cmd_classes in self.__commands_groups:
-            extra.append("    {}:".format(group_title))
-            for cmd_class in sorted(cmd_classes, key=operator.attrgetter('name')):
-                extra.append("        {:{longest}s}   {}".format(
-                    cmd_class.name, cmd_class.help_msg, longest=longest_name))
-            extra.append('')
+    def parse_known_args(self, args, namespace):
+        """Hold the namespace to verify if help was requested."""
+        if namespace is None:
+            namespace = argparse.Namespace()
+        self._namespace = namespace
+        return super().parse_known_args(args, namespace)
+
+    def _check_value(self, action, value):
+        """Verify the command is a valid one.
+
+        This overwrites ArgumentParser one to change the error text.
+        """
+        if action.choices is not None and value not in action.choices:
+            self._namespace._missing_command = True
+            raise argparse.ArgumentError(action, "no such command {!r}".format(value))
+
+    def error(self, message):
+        """Show the usage, the error message, and no more."""
+        # if help was requested, no matter other errors (just provide help!)
+        _missing_command = getattr(self._namespace, '_missing_command', False)
+        if _get_option(self._namespace, 'help') and not _missing_command:
+            help_text = get_help(self, self._namespace)
+            raise CommandError(help_text, argsparsing=True)
+
+        # the text for the usage is just 'charmcraft', unless we have a subparser in plase, so
+        # a command was used, so it's 'charmcraft <somecommand>'
+        fullcommand = self.prog
+        subparser = _get_subparser(self, self._namespace)
+        if subparser is not None:
+            fullcommand = subparser.prog
+
+        full_msg = helptexts.get_usage_message(fullcommand, message)
+        raise CommandError(full_msg, argsparsing=True)
+
+
+def _get_subparser(parser, namespace):
+    """Return a subparser for that command, if any."""
+    command = getattr(namespace, '_command', None)
+    if command is not None:
+        subactions = [
+            action for action in parser._actions
+            if isinstance(action, argparse._SubParsersAction)]
+        if subactions:
+            (subparseraction,) = subactions
```

Comment:
this kind of thing has bitten us before, where it explodes with an obscure error. Better to check the length and be explicit about what went wrong if it's not exactly 1, than this.

Reply:
Added a explicit check for that situation (but didn't add a test for that, as it would imply messing with argparse in a different way that we do now, kind of "breaking" the other part of the code)